PhysicsTools/PatAlgos/plugins/PATRefitVertexProducer.cc

```diff
@@ -0,0 +1,192 @@
+    if (key != vtxIdx ||
+        (quality != pat::PackedCandidate::UsedInFitTight && quality != pat::PackedCandidate::UsedInFitLoose))
+      continue;
+    if (useEleKfTracks_ && std::abs(cand.pdgId()) == 11)
+      continue;
+    transTracks.push_back(transTrackBuilder.build(cand.bestTrack()));
+  }
+
+  // loop over the lostTracks
+  if (useLostTracks_) {
+    for (const auto& cand : (*lostTrackCands)) {
+      if (cand.charge() == 0 || cand.vertexRef().isNull())
+        continue;
+      if (cand.bestTrack() == nullptr)
+        continue;
+      auto key = cand.vertexRef().key();
+      auto quality = cand.pvAssociationQuality();
+      if (key != vtxIdx ||
+          (quality != pat::PackedCandidate::UsedInFitTight && quality != pat::PackedCandidate::UsedInFitLoose))
+        continue;
+      transTracks.push_back(transTrackBuilder.build(cand.bestTrack()));
+    }
+  }
+
+  // loop over the electronKfTracks
+  if (useEleKfTracks_) {
+    for (const auto& cand : (*eleKfTrackCands)) {
+      if (cand.charge() == 0 || cand.vertexRef().isNull())
+        continue;
+      if (cand.bestTrack() == nullptr)
+        continue;
+      auto key = cand.vertexRef().key();
+      auto quality = cand.pvAssociationQuality();
+      if (key != vtxIdx ||
+          (quality != pat::PackedCandidate::UsedInFitTight && quality != pat::PackedCandidate::UsedInFitLoose))
+        continue;
+      transTracks.push_back(transTrackBuilder.build(cand.bestTrack()));
+    }
+  }
+
+  // Refit the vertex
+  TransientVertex transVtx;
+  reco::Vertex refitPV(pv);  // initialized to the original PV
+
+  bool fitOK = true;
+  if (transTracks.size() >= 3) {
+    AdaptiveVertexFitter avf;
+    avf.setWeightThreshold(0.1);  // weight per track, allow almost every fit, else --> exception
+    if (!useBeamSpot_) {
+      transVtx = avf.vertex(transTracks);
+    } else {
+      transVtx = avf.vertex(transTracks, *beamSpot);
+    }
+    if (!transVtx.isValid()) {
+      fitOK = false;
+    } else {
+      //MB: protect against rare cases when transVtx is valid but its postion is ill-defined
+      if (!std::isfinite(transVtx.position().z()))  //MB: it is enough to check one coordinate (?)
+        fitOK = false;
+    }
+  } else
+    fitOK = false;
+  if (fitOK) {
+    refitPV = transVtx;
+  }
+
+  outputVertices->push_back(refitPV);
+
+  evt.put(std::move(outputVertices));
+}
+
+void PATRefitVertexProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // patRefitVertexProducer
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("srcVertices", edm::InputTag("offlineSlimmedPrimaryVertices"));
+  desc.add<edm::InputTag>("srcCands", edm::InputTag("packedPFCandidates"));
+  desc.add<edm::InputTag>("srcLostTracks", edm::InputTag("lostTracks"));
+  desc.add<edm::InputTag>("srcEleKfTracks", edm::InputTag("lostTracks:eleTracks"));
+  desc.add<edm::InputTag>("srcBeamSpot", edm::InputTag("offlineBeamSpot"));
+  desc.add<bool>("useBeamSpot", true);
```

Comment:
how about a `setComment` for the non trivial parameters

Reply:
OK.